chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -189,4 +190,79 @@ Array Inverse(const Array& a) {
     return out;
 }
 
+std::tuple<Array, Array> Qr(const Array& a, QrMode mode) {
+    CheckRankTwoArray(a);
+    Device& device = a.device();
+    Dtype dtype = internal::GetMathResultDtype(a.dtype());
+
+    int64_t m = a.shape()[0];
+    int64_t n = a.shape()[1];
+    int64_t k = std::min(m, n);
+
+    Array tau = Empty(Shape{k}, dtype, device);
+    Shape q_shape{0};
+    Shape r_shape{0};
+    if (mode == QrMode::reduced) {
+        q_shape = Shape{m, k};
+        r_shape = Shape{k, n};
+    } else if (mode == QrMode::complete) {
+        q_shape = Shape{m, m};
+        r_shape = Shape{m, n};
+    } else if (mode == QrMode::r) {
+        r_shape = Shape{k, n};
+    } else if (mode == QrMode::raw) {
+        r_shape = Shape{n, m};  // for "raw" mode r in the code corrensponds to h in docs
+    }
```

Comment:
How about using switch cases here and in the Python bindings?

Reply:
Done here.
In Python bindings the conditional statements are with `mode` of type `std::string`, switch cases are not applicable there, right?